### PR TITLE
Ignore unsupported partition fields

### PR DESCRIPTION
**Pull request by fbocse**
Thought I'd give it a try and help resolve the forward-compatible issue for new transforms to the partition spec detailed on [issue#32](https://github.com/apache/incubator-iceberg/issues/32)

### CODE REVIEW

File: api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java [hunk 1]
```diff
@@ -80,4 +80,25 @@ public void testTransforms() throws Exception {
...
+        Types.NestedField.required(2, "d", Types.DateType.get()),
+        Types.NestedField.required(3, "s", Types.StringType.get())
+    );
+
+    PartitionSpec spec = PartitionSpec.builderFor(schema)
+        .identity("i")
+        .year("d")
+        .month("d")
+        .day("d")
+        .add(3, "b", "noSuchTransform?!")
```
Comment (rdblue) [inline api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java]: How does the builder succeed with this transform?

File: api/src/main/java/com/netflix/iceberg/transforms/Transforms.java [hunk 2]
```diff
@@ -41,27 +42,27 @@ private Transforms() {
...
     } else if (type.typeId() == Type.TypeID.TIMESTAMP) {
-      return Timestamps.valueOf(transform.toUpperCase(Locale.ENGLISH));
+      return Optional.of(Timestamps.valueOf(transform.toUpperCase(Locale.ENGLISH)));
     } else if (type.typeId() == Type.TypeID.DATE) {
-      return Dates.valueOf(transform.toUpperCase(Locale.ENGLISH));
+      return Optional.of(Dates.valueOf(transform.toUpperCase(Locale.ENGLISH)));
     }
 
-    throw new IllegalArgumentException("Unknown transform: " + transform);
+    return Optional.empty();
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/transforms/Transforms.java]: We don't use optional very much (if at all) in the Iceberg API. It is usually better to return something that behaves correctly and doesn't require the caller to handle a new case, like returning `ImmutableList.of()` for a list instead of `Optional.empty()`. When the caller should handle a case, we usually return `null` and document it to avoid instantiating extra objects. Both of these options don't require making type changes to the API.

In this case, I think the first option is a better implementation. This should return an `UnknownTransform` implementation. If a transform is not recognized, Iceberg can still read data, but cannot write data for that partition spec because it doesn't know how to derive a partition value using the transform. Returning `Optional.empty` here and ignoring unknown transforms in `PartitionSpec` would not signal that the partition spec can't be used to write. An older writer may try to add files and compact them with a manifest from a newer writer, which would cause problems.

An `UnknownTransform` implementation should throw an exception in `apply`. It should also return `Expressions.alwaysTrue` for `project` (all partition values may contain matching data) and `null` for `projectStrict` (no guarantees about matching a predicate).

**Comment by fbocse** [comment]
@rdblue I'm following up on your initial review and basically dropping the original change proposal, that was a bit naive. Since I believe that initially I've greatly underestimated the task at hand I wanted to ask whether you consider the current state of the PR as worth following up on?

File: api/src/main/java/com/netflix/iceberg/expressions/Expressions.java [hunk 3]
```diff
@@ -117,6 +117,10 @@ public static Expression not(Expression child) {
     return new UnboundPredicate<>(op, ref(name));
   }
 
+  public static <T> UnboundPredicate<T> alwaysTrue(String name) {
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/expressions/Expressions.java]: Why was this added? A `true` predicate has no field refence.

File: api/src/main/java/com/netflix/iceberg/transforms/Transforms.java [hunk 4]
```diff
@@ -56,12 +58,27 @@ private Transforms() {
...
+          () -> Dates.valueOf(transform.toUpperCase(Locale.ENGLISH)),
+          () -> new UnknownTransform<>(type, transform)
+      );
     }
 
-    throw new IllegalArgumentException("Unknown transform: " + transform);
+    return new UnknownTransform<>(type, transform);
+  }
+
+  private static Transform<?,?> match(Supplier<Transform> transformSupplier,
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/transforms/Transforms.java]: Nit: arguments should start at the same column or should be wrapped starting at 4 spaces from `private`.

File: api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+  @Override
+  public T apply(S value) {
+    throw new IllegalArgumentException(
+        "Unknown transform: " + transform + " for type: " + type.typeId().name());
+  }
+
+  @Override
+  public boolean canTransform(Type type) {
+    // Assuming the transform function can be applied for this type
+    return true;
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java]: I think it would be more clear to state that this assumes that the transform is valid because unknown transform is only used when parsing a transform in an existing table. That is, some other Iceberg version has already validated this. There is no way to use the unknown transform specifically.

File: api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+
+  @Override
+  public Type getResultType(Type sourceType) {
+    return type;
+  }
+
+  @Override
+  public UnboundPredicate<T> project(String name, BoundPredicate<S> predicate) {
+    // Assuming all partition values may contain matching data
+    return Expressions.alwaysTrue(name);
```
Comment (rdblue) [inline api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java]: I don't think passing the name is necessary.

File: api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java [hunk 7]
```diff
@@ -19,6 +19,7 @@
 
 package com.netflix.iceberg.transforms;
 
+
```
Comment (rdblue) [inline api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java]: Nit: please remove any independent whitespace changes to avoid commit conflicts.

File: api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java [hunk 8]
```diff
@@ -80,4 +81,19 @@ public void testTransforms() throws Exception {
     Assert.assertEquals("Deserialization should produce equal partition spec",
         spec, TestHelpers.roundTripSerialize(spec));
   }
+
+  @SuppressWarnings("unchecked")
+  @Test(expected = IllegalArgumentException.class)
+  public void testApplyUnknownTransform() {
```
Comment (rdblue) [inline api/src/test/java/com/netflix/iceberg/transforms/TestTransformSerialization.java]: Use `TestHelpers.assertThrows` instead of `expected`. It allows you to validate the exception message.

File: api/src/test/java/com/netflix/iceberg/transforms/TestUnknownTransform.java [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  @Test
+  public void testUnknownTransform() {
+    Schema schema = new Schema(
+        required(1, "id", Types.LongType.get()),
+        optional(2, "before", Types.DateType.get()),
+        required(3, "after", Types.DateType.get()));
+
+    PartitionSpec spec = PartitionSpec.builderFor(schema)
+        .day("before")
+        .add(3, "after_unknown_transform", "unknown")
```
Comment (rdblue) [inline api/src/test/java/com/netflix/iceberg/transforms/TestUnknownTransform.java]: I don't think that the builder should allow unknown transforms from the public API. This should go through a protected method.

File: api/src/main/java/com/netflix/iceberg/expressions/Expressions.java (hunk 3 above)
Comment (fbocse) [reply api/src/main/java/com/netflix/iceberg/expressions/Expressions.java]: I've added this method in `com.netflix.iceberg.expressions.Expressions` because `com.netflix.iceberg.expressions.UnboundPredicate` constructors are package-private so it's visible only within its own package and I needed to override method from class in different package `com.netflix.iceberg.transforms.UnknownTransform#project`. Does this answer your question? 
Since this method is obviously taking a redundant argument would you consider this version instead? 
```return new UnboundPredicate<>(Operation.TRUE, null);```

File: api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java (hunk 6 above)
Comment (fbocse) [reply api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java]: The effective return value of this method should be ```UnboundPredicate<>(Operation.TRUE, null)```, right? Given that `com.netflix.iceberg.expressions.UnboundPredicate` constructors are package private I thought I'd use `com.netflix.iceberg.expressions.Expressions` static methods (same package) to build UnboundPredicate instances. Does that sound reasonable or are you thinking of a better solution/refactoring?

File: api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java (hunk 5 above)
Comment (fbocse) [reply api/src/main/java/com/netflix/iceberg/transforms/UnknownTransform.java]: Thank you, will add this in, this definitely clears things for me too :)

File: api/src/test/java/com/netflix/iceberg/transforms/TestUnknownTransform.java (hunk 9 above)
Comment (fbocse) [reply api/src/test/java/com/netflix/iceberg/transforms/TestUnknownTransform.java]: I believe you are suggesting we keep `com.netflix.iceberg.PartitionSpec.Builder#add` throw `IllegalArgumentException` should `com.netflix.iceberg.transforms.Transforms#fromString` return `com.netflix.iceberg.transforms.UnknownTransform`? But Transforms it's also public API, right? So should we have separate protected methods in both the Builder and Transforms to accommodate unknown formats?
